% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/query_genespace.R
\name{query_genespace}
\alias{query_genespace}
\alias{query_hits}
\alias{query_pangenes}
\title{Query GENESPACE results}
\usage{
query_hits(gsParam, bed, synOnly = TRUE)

query_pangenes(
  gsParam,
  bed = NULL,
  refGenome = NULL,
  transform = TRUE,
  showArrayMem = TRUE,
  showNSOrtho = TRUE,
  maxMem2Show = Inf
)
}
\arguments{
\item{gsParam}{A list of genespace parameters created by init_genespace.}

\item{bed}{A data.table similar to the format of a bed file. At least two
columns must be specified: genome, chr. These must match genome-chromosome
combinations in the data. The user can also supply two additional columns:
start, end. These are the base-pair start and end coordinates of each region.
If start and end are not specified, they are set to 0 and Inf respectively
so that the entire chromosome is returned.}

\item{synOnly}{logical, given to query hits to return (or not) the synteny-
constrained hits only}

\item{transform}{logical, should the pangenome be transformed from the long
to the wide format?}

\item{showArrayMem}{logical, should all genes or only the array
representative genes (if FALSE) be shown?}

\item{showNSOrtho}{logical, should non-syntenic orthologs be included in the
output?}

\item{maxMem2Show}{integer specifying the maximum number of members to be
included in each entry-by-flag-by-genome combination.}
}
\value{
a list of data.tables named $genome, $chr: $start-$end. One
data.table for each line in the bed file.
}
\description{
\code{query_genespace} Functions to pull data stored in the /pangenome
and /syntenicHits directories

\code{query_hits} query_hits

\code{query_pangenome} query_pangenome
}
